File: python/longestPalindrome.py
'''
This solution works by expanding around the center to find palindromes

function LongestPalindrome
    if s is empty
    return

    longest = ""

    create loop from i to 0 to length of s - 1
        # check for odd length palidrome
        oddpalidrome = expandFunction(s,i,i)
        if length of oddplaidrome is greater than length of longest
            set longest as odd
        
        # check for even palidrome 
        evenpalidrome = expandFunction(s,i ,i)
        if length of even palidrome is greater than length of longest
            set longest as even

expandFunction(s, left, right)
    
    while left is >= 0 and right is < than length of s and s[left] = s[right]
        decrement left by one
        increment right by one
        
    return substring of s from left + 1 to right
    

'''

class Solution(object):
    # Definition that takes in self (to reference solution) and s (the input string)
    # First line is a conditional statement to check if the string is empty
    def longestPalindrome(self, s):
        if not s:
            return ""

        # Input string will be called longest
        longest = ""

        # Loop starts to iterate over the indices of string s
        # range(len(s)) creates a sequence of numbers from 0 to len(s) - 1; which are the valid indices in the string
        for i in range(len(s)):

            # Check for odd length palindromes
            # Calls expand_around_center method; passing s, i (as current index), and i (for center index)
            # The method results in the odd_palindrome- it does a quick check to see if it is the longest string
            odd_palindrome = self.expand_around_center(s, i, i)
            if len(odd_palindrome) > len(longest):
                longest = odd_palindrome

            # Check for even length palindromes
            # Same as above but this time the method expands from two adjacent indicies to find longest even length palidrome
            # Because it is even ;)
            even_palindrome = self.expand_around_center(s, i, i + 1)
            if len(even_palindrome) > len(longest):
                longest = even_palindrome

        return longest # longest palidrome should be returned

    def expand_around_center(self, s, left, right):
        # While loop will continue as three conditions are satisfied simulteanously: (ensuring they don't go out of bounds)
        # left: is great than or equal to 0 
        # right: less than the length of the string s (they also must be equal because palindrome)
        while left >= 0 and right < len(s) and s[left] == s[right]:

            # Left is decremented by 1 and right is incremented by one- palidrome range is expanded 
            left -= 1
            right += 1

        # Returns the SUBSTRING
        # This slices the s string to exclude the non-palindromic chars
        # Since the left will be the first to know of the bad character you must slice it using + 1
        return s[left + 1:right]

File: JAVA/longestPalindrome.java
class Solution {
    public String longestPalindrome(String s) {
        // Check if the string is null or has length less than 2, return the input string
        // Could just check if null
        if (s == null || s.length() < 2) {
            return s;
        }
        
        // Initialize start and end indices of the longest palindromic substring
        // IMPORTANT: using the start and end indices instead of saving the longest string as an array
        // is a memeory efficient approach. You don't need to store the substring- just indicate where it is
        int start = 0;
        int end = 0;
        
        // Iterate over each character in the string
        for (int i = 0; i < s.length(); i++) {
            // Expand around the center for odd-length palindromes
            int len1 = expandAroundCenter(s, i, i);
            
            // Expand around the center for even-length palindromes
            int len2 = expandAroundCenter(s, i, i + 1);
            
            // Compute the length of the longest palindrome at current center position
            int len = Math.max(len1, len2);
            
            // If the current palindrome is longer than the previously found longest palindrome,
            // update the start and end indices accordingly
            if (len > end - start) {
                start = i - (len - 1) / 2;
                end = i + len / 2;
            }
        }
        
        // Extract and return the longest palindromic substring using substring(start, end + 1)
        return s.substring(start, end + 1);
    }
    
    // Helper method to expand around the center of a palindrome
    private int expandAroundCenter(String s, int left, int right) {
        // Expand until the left and right indices are within bounds and characters are equal
        while (left >= 0 && right < s.length() && s.charAt(left) == s.charAt(right)) {
            left--;
            right++;
        }
        
        // Return the length of the palindrome (right - left - 1)
        return right - left - 1;
    }
}
